### PR TITLE
Allow setting the name for the base container within K8s Pod Operator

### PR DESCRIPTION
**Pull request by csm10495**
Some downstream machinery may require a specific name for the 'base' container.

This change should be backwards compatible since the default base container name is still base.. it can just now be changed to something besides base.

This is the second part of splitting up: #28752

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 1]
```diff
@@ -206,9 +206,10 @@ class KubernetesPodOperator(BaseOperator):
         to populate the environment variables with. The contents of the target
         ConfigMap's Data field will represent the key-value pairs as environment variables.
         Extends env_from.
+    :param base_container_name: The name of the base container in the pod. This container's logs
```
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:209]: Should we have some doc explnations about this? Mainly around what this capability brings

Comment (csm10495) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:209]: I don't think the concept of 'base container' is mentioned anywhere already. It seems to just be the name given to the first container.. and is also the one that can be logged via get_logs=True. That info is now in this docstr. Anywhere else where we should mention this?

**Review by ashb — APPROVED** [review]
LGTM, but I'll let Elad respond too.

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 2]
```diff
@@ -272,6 +273,7 @@ def __init__(
         pod_runtime_info_envs: list[k8s.V1EnvVar] | None = None,
         termination_grace_period: int | None = None,
         configmaps: list[str] | None = None,
+        base_container_name: str = "base",
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: I think we should switch this back so the default is still coming from `BASE_CONTAINER_NAME`. This makes it possible to change the default instance wide, and it also means this is a feature that can ship in a minor release vs being a breaking change that requires a major version.

Comment (csm10495) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: But what about for folks who don't want to change this instance wide and instead want to set it on a single basis? To fully maintain support for BASE_CONTAINER_NAME why not also have that and default this new param to None, then if this is set it takes precedence, if not falls back to BASE_CONTAINER_NAME?

Comment (jedcunningham) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: That's basically what I'm saying. The "how" (default of none vs the constant) doesn't really matter at the end of the day, they are functionally equivalent:

```
BASE_CONTAINER_NAME = "base"
...
base_container_name: str = BASE_CONTAINER_NAME
...
self.base_container_name = base_container_name
```

and 

```
BASE_CONTAINER_NAME = "base"
...
base_container_name: str | None = None
...
self.base_container_name = base_container_name or BASE_CONTAINER_NAME
```

Comment (csm10495) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Got it. Updated. Please check again.

**Comment by csm10495** [comment]
@jedcunningham check again when you can. Thanks!

**Comment by csm10495** [comment]
@eladkal can you check again?

**Review by potiuk — APPROVED** [review]
It's pretty clear to me (and used before in the code :D)

**Comment by jedcunningham** [comment]
Looking at this with fresh eyes, this works?

```
kpo_task = KubernetesPodOperator(...)
kpo_task.BASE_CONTAINER_NAME = "whatever"
```

Is it worth adding this to init for an edge case?

**Comment by csm10495** [comment]
It doesn't really match how params are normally passed through.

Also it would look kind of weird in a taskflow dag. I think the PR as is makes a bit more sense.

**Comment by csm10495** [comment]
Also does that change it just for the instance or for the class if done that way? .. I'd have to check to know for sure.

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 3]
```diff
@@ -1153,3 +1153,134 @@ def test_using_resources(self):
...
+        assert mock_fetch_container_logs.call_args[1]["container_name"] == "apple-sauce"
+        actual_pod = self.api_client.sanitize_for_serialization(k.pod)
+        self.expected_pod["spec"]["containers"][0]["name"] = "apple-sauce"
+        assert self.expected_pod["spec"] == actual_pod["spec"]
+
+    def test_changing_base_container_name_no_logs(self):
+        """
+        This test checks BOTH a modified base container name AND the get_logs=False flow..
+          and as a result, also checks that the flow works with fast containers
+          See #26796
```
Comment (jedcunningham) [inline kubernetes_tests/test_kubernetes_pod_operator.py]:
```suggestion
        This test checks BOTH a modified base container name AND the get_logs=False flow,
        and as a result, also checks that the flow works with fast containers
        See https://github.com/apache/airflow/issues/26796
```

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 4]
```diff
@@ -1153,3 +1153,134 @@ def test_using_resources(self):
...
+
+        assert mock_await_container_completion.call_args[1]["container_name"] == "apple-sauce"
+        actual_pod = self.api_client.sanitize_for_serialization(k.pod)
+        self.expected_pod["spec"]["containers"][0]["name"] = "apple-sauce"
+        assert self.expected_pod["spec"] == actual_pod["spec"]
+
+    def test_changing_base_container_name_no_logs_long(self):
+        """
+        Similar to test_changing_base_container_name_no_logs, but ensures that
+        pods running longer than 1 second work too. See #26796
```
Comment (jedcunningham) [inline kubernetes_tests/test_kubernetes_pod_operator.py]:
```suggestion
        pods running longer than 1 second work too.
        See https://github.com/apache/airflow/issues/26796
```

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**